cocotb/types/array.py

```diff
@@ -0,0 +1,233 @@
+        Array(['1', '2', '3', '4'], Range(0, 3))
+
+        >>> Array(range=Range(0, 'downto', -3))  # the initial values are `None`
+        Array([None, None, None, None], Range(0, -3))
+
+        >>> Array([1, True, object(), 'example'], Range(-2, 1))  # inital value and range lengths must be equal
+        Array([1, True, <object object at 0x7f4cff7b5570>, 'example'], Range(-2, 'to', 1))
+
+    Arrays also support "null" ranges; "null" arrays have zero length and cannot be indexed.
+
+    .. code-block:: python3
+
+        >>> Array(range=Range(0, 'to', 1))
+        Array([], Range(1, 'to', 0))
+
+    Indexing and slicing is very similar to :class:`list`\ s, except you use the indexing scheme you specified.
+    Like :class:`list`\ s, you don't have to specify a start or stop, and the start or end of the array are infered.
+    Slicing an array returns a new :class:`~cocotb.types.Array` object, whose bounds are the slice indexes.
+
+    .. code-block:: python3
+
+        >>> a = Array("1234abcd")
+        >>> a[7]
+        'd'
+        >>> a[2:5]
+        Array(['3', '4', 'a', 'b'], Range(2, 'to', 5))
+        >>> a[2:5] = reversed(a[2:5])
+        >>> ''.join(a)
+        '12ba43cd'
+
+        >>> b = Array("1234", Range(0, 'downto', -3))
+        >>> b[-2]
+        '3'
+        >>> b[-1:]
+        Array(['2', '3', '4'], Range(-1, 'downto', -3))
+        >>> b[:] = reversed(b)
+        >>> b
+        Array(['4', '3', '2', '1'], Range(0, 'downto', -3))
+
+    .. note::
+        Slice indexes must be specified in the same direction as the array and do not support specifying a "step".
+
+    .. note::
+        When setting a slice, the new value must be an iterable the same size as the slice you are selecting.
+
+    .. note::
+        Negative indexes are *not* treated as an offset from the end of the array, but are treated literally.
+
+    Arrays support the methods and semantics defined by :class:`collections.abc.Sequence` including, but not limited to:
+
+    - ``len(array)`` for getting the length of the array,
+    - ``value in array`` to check if an array contains a value,
+    - ``array.index(value)`` to find the index of the first occurences of a value in the array.
+
+    Args
+        value: Initial value for the array.
+        range: Indexing scheme of the array.
+    """
+
+    __slots__ = ("_value", "_range", "__dict__")  # __dict__ necessary for cached_property
+
+    def __init__(
+        self, value: Optional[Iterable[Any]] = None, range: Optional[Range] = None
+    ):
+        if value is not None and range is None:
+            self._value = list(value)
+            self._range = Range(0, "to", len(self._value) - 1)
+        elif value is not None and range is not None:
+            self._value = list(value)
+            self._range = range
+            if len(self._value) != len(self._range):
+                raise ValueError("Init value does not fit in given range")
```

Comment:
Being more verbose might help debugging later.

Reply:
It would, but despite that most of the built in types have concise and low detail error message. I'm guessing the thought is you should use a debugger? I'll dive more into that since I haven't seen much opinion one way or the other.

Reply:
I googled and asked a few people and there is no clear guidance. I'll add the slice value and the length of the value to the error message. I don't the value to be printed since it could be rather large.